pulsar-client/src/test/java/org/apache/pulsar/client/impl/PartitionedProducerImplTest.java

```diff
@@ -264,4 +266,49 @@ public void testGetNumOfPartitions() throws Exception {
         assertEquals(producerImpl.getNumOfPartitions(), 0);
     }
 
+    @Test
+    public void testFlushWhenLastSendFutureFailed() {
+        ProducerConfigurationData producerConfData = new ProducerConfigurationData();
+        ProducerImpl<Object> producerImpl = client.newProducerImpl(TOPIC_NAME, 0, producerConfData,
+                null, null, null, Optional.empty());
+
+        // 1. When no data is sent to this producerImpl,
+        // its lastSendFuture is always in normal completion state
+        CompletableFuture<Void> lastSendFuture = producerImpl.getLastSendFuture();
+        assertTrue(lastSendFuture.isDone());
+        assertFalse(lastSendFuture.isCompletedExceptionally());
+
+        // 2. We set the lastSendFuture of this partition to an abnormal state,
+        // simulating that an exception occurred during the sending process
+        final String failedMessage = "failed last send future";
+        producerImpl.setLastSendFuture(FutureUtil.failedFuture(new Throwable(failedMessage)));
+
+        // 3. So when we get its lastSendFuture again,
+        // the future is already in an abnormal failure state
+        lastSendFuture = producerImpl.getLastSendFuture();
+        assertTrue(lastSendFuture.isDone());
+        assertTrue(lastSendFuture.isCompletedExceptionally());
+        // 4. The following simple simulation application captures the exception processing
+        lastSendFuture.exceptionally(throwable -> {
+            assertNotNull(throwable);
+            assertEquals(throwable.getMessage(), failedMessage);
+            return null;
+        });
+
+        // 5. We have already handled the exception in the step 4,
+        // and then PartitionedProducerImpl will continue to send data.
+        // It should be noted that when this partition is not selected for data transmission this time,
+        // its lastSendFuture is still the future that was in an abnormal state in the previous step.
+        // When the application calls the flush operation again, in the previous logic,
+        // its exception future will be returned to the application,
+        // causing the application to always execute the exception handling logic.
+        // In fact, we have handled the exception before,
+        // and we did not send data to this partition this time,
+        // it should not affect this transmission.
+        // So we want the lastSendFuture here to be in normal state.
+        lastSendFuture = producerImpl.getLastSendFuture();
+        assertTrue(lastSendFuture.isDone());
```

Comment:
this should be tested by observing the external behavior. for example, sending a message while the connection is broken. the flush should fail only once and it should be possible to send more messages after the connection resumes.

Reply:
I agree with you, I'm not very familiar with pulsar's testing now, I need to study how to disconnect and reconnect again during the process of sending data to see if it works as expected.

Reply:
@wenbingshen - let me know if you need pointers to relevant tests. I can try to help find good tests tomorrow (my time).

Reply:
@michaeljmarshall It would be great if you could help collaborate on this PR. As this issue is happening in our online environment and I hope it gets verified and resolved soon. Thanks. :)

Reply:
@wenbingshen Here are examples of testing the external behavior instead of observing internal state: https://github.com/apache/pulsar/blob/22a0beb5d4dc4c57fede791194573535bc7b8b10/pulsar-broker/src/test/java/org/apache/pulsar/client/api/SimpleProducerConsumerTest.java#L572-L607

Most Pulsar broker tests are more or less integration tests. The problem with Pulsar coding style is that it's not very modular. Class based unit tests aren't really meaningful since the class is not a modular unit in most cases. A meaningful unit is hardly ever a single class and it is considered as an anti-pattern to have class based unit tests since it leads to the bad test design and tests that are a maintenance burden (references: [Gulliver's Travels Tests](https://dhickey.ie/2014/03/03/gullivers-travels-test/), [What is a unit test](https://tanzu.vmware.com/content/blog/what-is-a-unit-test-the-answer-might-surprise-you))

With modularity I mean in this case "units" (not necessarily a single class) would have a well defined public interface. In unit testing, only public interfaces should be tested and outputs should be asserted. "output" could be also calls to other collaborators. Different developers have different preferences of unit testing styles and there are multiple schools of how to do unit testing (for example ["Detroit and London Schools of TDD"](https://blog.devgenius.io/detroit-and-london-schools-of-test-driven-development-3d2f8dca71e5)) . In Pulsar, we haven't had these deep conversations about the styles for TDD and how to improve the situation.

I think that integration tests have more value than tests that observe internal private state. Once you let go of naming the test class to match the implementation class, it will be easier to name the test class. In Pulsar, the "SimpleProducerConsumerTest" class is a big container for a lot of (integration) tests. That style isn't optimal either. I think it's better to create more test classes that are categorized by behavior or scenarios instead of dumping everything in "SimpleProducerConsumerTest". There's already the "ProducerConsumerBase" base class and creating new test classes is easy. In some ways, "TopicDuplicationTest" is a good example where certain scenarios are in the same test class instead of mixing a lot of different type of scenarios as there is in SimpleProducerConsumerTest . I hope there's a useful starting point based on this information.

Reply:
> this should be tested by observing the external behavior

+1 to this (thanks @lhotari )

Reply:
> > this should be tested by observing the external behavior
> 
> +1 to this (thanks @lhotari )

+1

Also, thanks @lhotari for adding test examples.